File: haskell/list-ops/ListOps.hs
{-# LANGUAGE BangPatterns #-}

module ListOps
  ( length
  , reverse
  , map
  , filter
  , foldr
  , foldl'
  , (++)
  , concat
  ) where

import Prelude hiding
  ( length, reverse, map, filter, foldr, (++), concat )

foldl' :: (b -> a -> b) -> b -> [a] -> b
foldl' _ !acc [] = acc
foldl' f !acc (x:xs) = foldl' f (f acc x) xs

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr _ acc [] = acc
foldr f acc (x:xs) = f x $ foldr f acc xs

length :: [a] -> Int
length = foldl' (\x y -> x+1) 0

reverse :: [a] -> [a]
reverse = foldl' (\x y -> y:x) []

map :: (a -> b) -> [a] -> [b]
map _ [] = []
map f (x:xs) = f x : map f xs

filter :: (a -> Bool) -> [a] -> [a]
filter f [] = []
filter f (x:xs)
        | f x = x : filter f xs
        | otherwise = filter f xs

(++) :: [a] -> [a] -> [a]
xs ++ ys = foldr (:) ys xs

concat :: [[a]] -> [a]
concat = foldr (++) []

File: go/list-ops/listops.go
package listops

type IntList []int
type binFunc func(int, int) int

func (il IntList) Foldr(fn binFunc, zero int) int {
	if len(il) == 0 {
		return zero
	}
	total := fn(il[len(il)-1], zero)
	for i := len(il) - 2; i >= 0; i-- {
		total = fn(il[i], total)
	}

	return total
}

func (il IntList) Foldl(fn binFunc, zero int) int {
	if len(il) == 0 {
		return zero
	}

	total := fn(zero, il[0])
	for _, i := range il[1:] {
		total = fn(total, i)
	}

	return total
}

type predFunc func(a int) bool

func (il IntList) Filter(pf predFunc) IntList {
	res := IntList{}
	for _, i := range il {
		if pf(i) {
			res = append(res, i)
		}
	}
	return res
}

func (il IntList) Length() int {
	return len(il)
}

type unaryFunc func(i int) int

func (il IntList) Map(fn unaryFunc) IntList {
	res := IntList{}
	for _, i := range il {
		res = append(res, fn(i))
	}

	return res
}

func (il IntList) Reverse() IntList {
	res := IntList{}
	for i := len(il) - 1; i >= 0; i-- {
		res = append(res, il[i])
	}
	return res
}

func (il IntList) Append(i IntList) IntList {
	if len(il) == 0 {
		return i
	}

	return append(il, i...)
}

func (il IntList) Concat(ls []IntList) IntList {
	res := il
	for _, l := range ls {
		res = append(res, l...)
	}

	return res
}
